doc: add migration guide (#653)
operator-framework/operator-sdk: doc/migration/v0.1.0-migration-guide.md

# Migration Guide from v0.0.x to v0.1.0

This document describes how to migrate an operator project from Operator SDK `v0.0.x` to `v0.1.0`. For an overview of the major changes in the `v0.1.0` project layout and APIs see the [v0.1.0 changes doc][v0.1.0-changes-doc].

> The recommended way to migrate your project is to initialize a new `v0.1.0` project and copy and modify your code into the new project.

Reviewer: ```suggestion
The recommended way to migrate your project is to initialize a new `v0.1.0` project, then copy your code into the new project and modify as described below.
```

> This guide goes over the example of migrating the memcached-operator from the user guide to illustrate the migration steps. See the example [v0.0.7 memcached-operator][v0.0.7-memcached-operator] for the project before the migration and [v0.1.0 memcached-operator][v0.1.0-memcached-operator] after migration.

Reviewer: ```suggestion
This guide goes over migrating the memcached-operator, an example project from the user guide, to illustrate migration steps. See the [v0.0.7 memcached-operator][v0.0.7-memcached-operator] and [v0.1.0 memcached-operator][v0.1.0-memcached-operator] project structures for pre- and post-migration examples, respectively.
```

> This guide goes over migrating the memcached-operator, an example project from the user guide, to illustrate migration steps. See the [v0.0.7 memcached-operator][v0.0.7-memcached-operator] and [v0.1.0 memcached-operator][v0.1.0-memcached-operator] project structures for pre- and post-migration examples, respectively.

Author: Tags have a small issue that while the memcached-operator example is updated on the master, the other example operators like the vault-operator and the ansible operators are not.
So it's hard to tag that as `v0.1.0`. Same applies for the branch.

We need to get better at keeping all the examples updated and tagging them as we release the SDK. If we tag the samples repo for `v0.1.0` then I'll update this later. For now though I think linking to the commits should be enough to demonstrate the example.

```Go
import (
    cachev1alpha1 "github.com/example-inc/memcached-operator/pkg/apis/cache/v1alpha1"
    ...
)

func add(mgr manager.Manager, r reconcile.Reconciler) error {
    c, err := controller.New("memcached-controller", mgr, controller.Options{Reconciler: r})

    // Watch for changes to the primary resource Memcached
    err = c.Watch(&source.Kind{Type: &cachev1alpha1.Memcached{}}, &handler.EnqueueRequestForObject{})

    // Watch for changes to the secondary resource Pods and enqueue reconcile requests for the owner Memcached
    err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
		IsController: true,
		OwnerType:    &cachev1alpha1.Memcached{},
	})
}
```

Watching multiple resources lets you trigger the reconcile loop for multiple resources relevant to your application. See the [watching and eventhandling][watching-eventhandling-doc] doc and the Kubernetes [controller conventions][controller-conventions] doc for more details. 

Remove the second `Watch()` or modify it to watch a secondary resource type that is owned by your CR.

#### Multiple custom resources

> If your operator is watching more than 1 CR type then you can do one of the following depending on you application:

Reviewer 2: ```suggestion
If your operator is watching more than 1 CR type then you can do one of the following depending on your application:
```

Copy the code from the old project's `Handle()` function over the existing code in your controller's `Reconcile()` function.
Be sure to keep the initial section in the `Reconcile()` code that looks up the object for the Request and checks to see if it's deleted.

```Go
import (
    apierrors "k8s.io/apimachinery/pkg/api/errors"
    cachev1alpha1 "github.com/example-inc/memcached-operator/pkg/apis/cache/v1alpha1"
    ...
)
func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Result, error) {
	// Fetch the Memcached instance
	instance := &cachev1alpha1.Memcached{}
	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
	if err != nil {
		if apierrors.IsNotFound(err) {
			// Request object not found, could have been deleted after reconcile request.
			// Owned objects are automatically garbage collected.
			// Return and don't requeue
			return reconcile.Result{}, nil
		}
		// Error reading the object - requeue the request.
		return reconcile.Result{}, err
    }

    // Rest of your reconcile code goes here.
    ...
}
```

Change the following in your reconcile code:
- Replace `return err` with `return reconcile.Result{}, err`

> - Replace the calls to the sdk client(Create, Update, Delete, Get, List) with the reconciler's client

Reviewer: ```suggestion
- Replace the calls to the SDK (Create, Update, Delete, Get, List) with the Reconciler's `client`
```

Author: This should still be `SDK client`. But I'll capitalize the SDK.

See the examples below and the controller-runtime [client API doc][client-api-doc] for more details.
```Go

> // Create

Reviewer 2: Just a suggestion, but for easier reading maybe a table would be better:

|        | v0.0.x                                                          | v.0.1                                                 |
|--------|-----------------------------------------------------------------|-------------------------------------------------------|
| Create | ```Go dep := &appsv1.Deployment{...} err := sdk.Create(dep) ``` | ```Go err := r.client.Create(context.TODO(), dep) ``` |
|   ...     |                              ...                                   |                       ...                                |
|  ...      |                         ...                |                                ...                       |

Author: @LiliC I've formatted it as a table and split the code into multi-line rows, but the lines wrap around a bit when I view the full doc. Not sure if that makes it as clear. Plus I can't seem to format the single lines as Go code. Seems like it has to be a block. You can see that in your comment above as well.

I've left both the table and code block in for comparison so you can see.
Also a screenshot of what I'm seeing:
![image](https://user-images.githubusercontent.com/6660644/47385697-99aab800-d6bf-11e8-94b8-790f947c491c.png)

Author: I'll wait to see if we can't format the table a bit better. Else I'm leaning towards keeping the code block which is a little more readable.

Reviewer 2: Makes sense 👍

dep := &appsv1.Deployment{...}
err := sdk.Create(dep)
// v0.0.1
err := r.client.Create(context.TODO(), dep)

// Update
err := sdk.Update(dep)
// v0.0.1
err := r.client.Update(context.TODO(), dep)

// Delete
err := sdk.Delete(dep)
// v0.0.1
err := r.client.Delete(context.TODO(), dep)

// List
podList := &corev1.PodList{}

> listOps := &metav1.ListOptions{LabelSelector: labelSelector}

Reviewer 3: `labelSelector` hasn't been defined at this point; you should move the label selector declaration before this line

err := sdk.List(memcached.Namespace, podList, sdk.WithListOptions(listOps))
// v0.1.0
labelSelector := labels.SelectorFromSet(labelsForMemcached(memcached.Name))
listOps := &client.ListOptions{Namespace: memcached.Namespace, LabelSelector: labelSelector}
err := r.client.List(context.TODO(), listOps, podList)

// Get
dep := &appsv1.Deployment{APIVersion: "apps/v1", Kind: "Deployment", Name: name, Namespace: namespace}
err := sdk.Get(dep)
// v0.1.0
dep := &appsv1.Deployment{}
err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, dep)
```

> Lastly copy and intialize any other fields that you may have had in your `Handler` struct into the `Reconcile<Kind>` struct:

Reviewer 2: s/intialize/initialize

```Go
// newReconciler returns a new reconcile.Reconciler

> func newReconciler(mgr manager.Manager) reconcile.Reconciler {

Author: @shawn-hurley dependency injection for the reconciler? Is that how we initialize the client with the manager's client? That's already shown as `&ReconcileMemcached{client: mgr.GetClient(), ...}`.
Or did you mean something else.

> func newReconciler(mgr manager.Manager) reconcile.Reconciler {

Author: Yeah we can go over that in a separate doc. For the migration users most likely won't need to inject fields into the controller, since the sdk handler should map directly onto the reconciler.

	return &ReconcileMemcached{client: mgr.GetClient(), scheme: mgr.GetScheme(), foo: "bar"}
}

// ReconcileMemcached reconciles a Memcached object
type ReconcileMemcached struct {
    client client.Client
    scheme *runtime.Scheme
    // Other fields
    foo string
}
```

### Copy changes from main.go

The main function for a `v0.1.0` operator in `cmd/manager/main.go` sets up the [Manager][manager-go-doc] which registers the custom resources and starts all the controllers.

> There is no need to migrate the sdk functions `sdk.Watch()`,`sdk.Handle()`, and `sdk.Run()` from the old `main.go` since that logic is now defined in a controller.

Reviewer: Upper-case 'sdk': ... SDK functions ...

However if there are any operator specific flags or settings defined in the old main file copy those over.

> If you have any 3rd party resource types registered with the sdk's scheme, then register those with the manager's scheme in the new project. See how to [register 3rd party resources][register-3rd-party-resources].

Reviewer: Upper-case 'sdk'

### Copy user defined files

If there are any user defined pkgs, scripts, and docs in the older project, copy these files into the new project. 

> ### Copy changes to deployment manifests

Reviewer 2: Nit: New line below missing.

For any updates made to the following manifests in the old project, copy over the changes to their corresponding files in the new project:

> - `tmp/build/Dockerfile` to `build/Dockerfile`

Reviewer 4: mention that the `tmp` directory no longer exists?

> - RBAC rules updates from `deploy/rbac.yaml` to `deploy/role.yaml`

Reviewer 3: ```suggestion
- RBAC rules updates from `deploy/rbac.yaml` to `deploy/role.yaml` and `deploy/role_binding.yaml`
```

- `deploy/cr.yaml` to `deploy/crds/<group>_<version>_<kind>_cr.yaml`
- `deploy/crd.yaml` to `deploy/crds/<group>_<version>_<kind>_crd.yaml`

### Copy user defined dependencies

For any user defined dependencies added to the old project's Gopkg.toml, copy and append them to the new project's Gopkg.toml.
Run `dep ensure` to update the vendor in the new project.

### Confirmation

At this point you should be able to build and run your operator to verify that it works. See the [user-guide][user-guide-build-run] on how to build and run your operator.

[v0.1.0-changes-doc]: ./v0.1.0-changes.md
[v0.0.7-memcached-operator]: https://github.com/operator-framework/operator-sdk-samples/tree/aa15bd278eec0959595e0a0a7282a26055d7f9d6/memcached-operator

> [v0.1.0-memcached-operator]: https://github.com/operator-framework/operator-sdk-samples/tree/master/memcached-operator

Reviewer 3: This should probably be locked to a specific commit so this doc stays consistent as we update the SDK over time

# Migration Guide from v0.0.x to v0.1.0

This document describes how to migrate an operator project built using Operator SDK `v0.0.x` to the project structure required by `v0.1.0`. 

The recommended way to migrate your project is to initialize a new `v0.1.0` project, then copy your code into the new project and modify as described below.

This guide goes over migrating the memcached-operator, an example project from the user guide, to illustrate migration steps. See the [v0.0.7 memcached-operator][v0.0.7-memcached-operator] and [v0.1.0 memcached-operator][v0.1.0-memcached-operator] project structures for pre- and post-migration examples, respectively.

## Create a new v0.1.0 project

Rename your `v0.0.x` project and create a new `v0.1.0` project in its place.

```sh
# Ensure SDK version is v0.1.0
$ operator-sdk --version
operator-sdk version 0.1.0

# Create new project
$ cd $GOPATH/src/github.com/example-inc/
$ mv memcached-operator old-memcached-operator
$ operator-sdk new memcached-operator
$ ls
memcached-operator old-memcached-operator

# Copy over .git from old project

> $ cp -rf memcached-operator/.git old-memcached-operator/.git 

Reviewer 5: This copies new to old, right? Don't you want this:
```
cp -rf old-memcached-operator/.git memcached-operator/.git
```

Author: Yes it should be the other way around, good catch.

Copy the code from the old project's `Handle()` function over the existing code in your controller's `Reconcile()` function.
Be sure to keep the initial section in the `Reconcile()` code that looks up the object for the Request and checks to see if it's deleted.

```Go
import (
    apierrors "k8s.io/apimachinery/pkg/api/errors"
    cachev1alpha1 "github.com/example-inc/memcached-operator/pkg/apis/cache/v1alpha1"
    ...
)
func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Result, error) {
	// Fetch the Memcached instance
	instance := &cachev1alpha1.Memcached{}
	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
	if err != nil {
		if apierrors.IsNotFound(err) {
			// Request object not found, could have been deleted after reconcile request.
			// Owned objects are automatically garbage collected.
			// Return and don't requeue
			return reconcile.Result{}, nil
		}
		// Error reading the object - requeue the request.
		return reconcile.Result{}, err

>     }

Reviewer 5: nit: this closing bracket (and the below comments) should be indented at the same level as `if err != nil {` in line 153

    // Rest of your reconcile code goes here.
    ...
}
```

Change the following in your reconcile code:
- Replace `return err` with `return reconcile.Result{}, err`
- Replace the calls to the SDK client(Create, Update, Delete, Get, List) with the reconciler's client

> 

Reviewer 5: Do we want to add:
```md
- Replace `return nil` with `return reconcile.Result{}, nil`
```
Should we mention `RequeueAfter`? We allude to this possibility a few paragraphs up, but since we call out how to change `return err`, we may want to be more explicit for `return nil` as well?

Author: Yes I'll mention how to replace `return nil` as well.

And I'll add a small section on periodic reconcile and `RequeueAfter`. We've mentioned it in the user-guide but it isn't obvious how this replaces the way we used to do periodic reconcile in the SDK.
